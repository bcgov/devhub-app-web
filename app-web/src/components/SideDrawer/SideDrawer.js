/*
Copyright 2019 Province of British Columbia

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at 

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Created by Patrick Simonian
*/
import React from 'react';
import PropTypes from 'prop-types';
import Aux from '../../hoc/auxillary';
import styled from '@emotion/styled';
import { css } from '@emotion/core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faTimes } from '@fortawesome/free-solid-svg-icons';
import Backdrop from '../UI/Backdrop/Backdrop';

const withPadding = css`
  padding: 10px;
`;

const Div = styled.div`
  position: fixed;
  top: 0;
  bottom: 0;
  width: 275px;
  background-color: #fff;
  z-index: 300;
  border-right: 1px solid #ccc;
  display: flex;
  flex-flow: column nowrap;
`;

const H2 = styled.h2`
  ${withPadding}
  justify-content: space-between;
  align-items: center;
  padding: 10px 5px;
  display: flex;
  color: #444;
  margin-bottom: 5px;
  flex: 0 0 auto;
  border-bottom: 1px solid #ccc;
`;

const IconButton = styled.button`
  font-size: 1.5em;
  margin-left: 5px;
  align-self: flex-start;
  cursor: pointer;
  border: none;
  background: transparent;
  :focus {
    outline: none;
  }
`;

const Menu = styled.div`
  ${withPadding}
  overflow: auto;
  flex-grow: 1;
`;

const SideDrawer = ({ closeDrawer, children, title, show }) =>
  show && (
    <Aux>
      <Backdrop clicked={closeDrawer} />
      <Div>
        <H2>
          <span>{title}</span>
          <IconButton onClick={closeDrawer}>
            <FontAwesomeIcon icon={faTimes} />
          </IconButton>
        </H2>
        <Menu>{children}</Menu>
      </Div>
    </Aux>
  );

SideDrawer.propTypes = {
  title: PropTypes.string,
  children: PropTypes.node,
  closeDrawer: PropTypes.func.isRequired,
  show: PropTypes.bool,
};

SideDrawer.defaultProps = {
  title: '',
  children: null,
  show: false,
};

export default SideDrawer;
