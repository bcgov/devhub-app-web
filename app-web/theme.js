/*
Copyright 2019 Province of British Columbia

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at 

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Created by Patrick Simonian
*/

/*
Emotion theming object

We are transitioning from common elements such as colors, background colors being abstracted away into
a themes provider. This is a very common practice now a days and makes it maintain themes on the fly
docs on theming is here https://emotion.sh/docs/theming#___gatsby

Docs on emotion https://emotion.sh/docs/introduction

--- our idea is to introduce emotion early and slowly replace css modules with emotion styled components
when and as needed. This is a slow processes but will lead to better and more maintaible styling in the long run.
*/

import { SPACING } from './src/constants/designTokens';
import { RESOURCE_TYPES, TOPICS, SEARCH_RESOURCE_TYPES, JOURNEY } from './src/constants/ui';

export const DEVHUB_PALETTE = {
  [RESOURCE_TYPES.COMPONENTS]: '#4299D2',
  [RESOURCE_TYPES.DOCUMENTATION]: '#246BD1',
  [RESOURCE_TYPES.PEOPLE]: '#E17039',
  [RESOURCE_TYPES.SELF_SERVICE_TOOLS]: '#F3BA45',
  [TOPICS]: '#444',
  [RESOURCE_TYPES.REPOSITORIES]: '#6e5494',
  [SEARCH_RESOURCE_TYPES.GITHUB_ISSUE]: '#6e5494',
  [RESOURCE_TYPES.EVENTS]: '#f44b42',
  [JOURNEY]: '#009688',
  lightgreen: '#66bb6a',
  yellow: '#FCBA19',
  lightyellow: '#F3BA45',
  grey: '#494949',
  lightgrey: '#F2F2F2',
  darkgrey: '#444',
  orange: '#E17039',
  purple: '#6E5494',
  red: '#D8292F',
  lightred: '#F44B42',
  green: '#2E8540',
  blue: '#003366',
  babyblue: '#4299D2',
  lightblue: '#246BD1',
  linkblue: '#1A5A96',
  bgBlue: '#38598A',
  white: '#ffffff',
};

export default {
  colors: {
    ...DEVHUB_PALETTE,
  },
  padding: {
    main: {
      mobile: `${SPACING['2x']} ${SPACING['2x']}`,
      desktop: `${SPACING['2x']} ${SPACING['13x']}`,
    },
  },
  breakpoints: {
    // specific breakpoints
    main: {
      desktop: '@media screen and (min-width: 932px)',
    },
  },
};
