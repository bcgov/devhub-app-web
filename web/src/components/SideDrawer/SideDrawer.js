/*
Copyright 2019 Province of British Columbia

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at 

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Created by Patrick Simonian
*/
import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { css } from '@emotion/react';
import posed from 'react-pose';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faTimes } from '@fortawesome/free-solid-svg-icons';
import Backdrop from '../UI/Backdrop/Backdrop';

const withPadding = css`
  padding: 10px;
`;

// create a pose component that can handle animating
const SlideInLeft = posed.div({
  show: {
    x: 0,
    transition: {
      duration: 300,
    },
  },
  hide: {
    x: '-100%',
    transition: {
      duration: 300,
    },
  },
});

// fit posed component into styled component for further styling
const Drawer = styled(SlideInLeft)`
  position: fixed;
  top: 0;
  bottom: 0;
  width: 275px;
  background-color: #fff;
  z-index: 300;
  border-right: 1px solid #ccc;
  display: flex;
  transform: translateX(-100%);
  flex-flow: column nowrap;
`;

const H2 = styled.h2`
  ${withPadding}
  justify-content: space-between;
  word-break: break-word;
  font-size: 25px;
  align-items: center;
  padding: 10px 5px;
  display: flex;
  color: #444;
  margin-bottom: 5px;
  margin-right: 30px;
  padding-right: 30px;
  flex: 0 0 auto;
  border-bottom: 1px solid #ccc;
`;

const TitleDiv = styled.div`
  display: inline-block;
  bottom-margin: 0px;
`;

const IconButton = styled.button`
  font-size: 1em;
  position: absolute;
  right: 5px;
  margin-left: 10px;
  align-self: flex-start;
  cursor: pointer;
  border: none;
  background: transparent;
  :focus {
    outline: none;
  }
`;

const Menu = styled.div`
  ${withPadding}
  overflow: auto;
  flex-grow: 1;
`;

export const TEST_IDS = {
  drawer: 'side-drawer',
  closeButton: 'side-drawer-close-button',
  backdrop: 'side-drawer-backdrop',
};
export const SideDrawer = ({ closeDrawer, children, title, show, addContent }) => {
  let content = '';
  if (addContent) {
    content = 'Content';
  }
  return (
    <React.Fragment>
      <Backdrop clicked={closeDrawer} show={show} data-testid={TEST_IDS.backdrop} />
      <Drawer pose={show ? 'show' : 'hide'} data-testid={TEST_IDS.drawer}>
        <H2>
          <TitleDiv>
            <span>{title}</span>
            <br />
            {content}
          </TitleDiv>
          <IconButton onClick={closeDrawer} data-testid={TEST_IDS.closeButton}>
            <FontAwesomeIcon icon={faTimes} />
          </IconButton>
        </H2>
        <Menu>{children}</Menu>
      </Drawer>
    </React.Fragment>
  );
};

SideDrawer.propTypes = {
  title: PropTypes.string,
  children: PropTypes.node,
  closeDrawer: PropTypes.func.isRequired,
  show: PropTypes.bool,
};

SideDrawer.defaultProps = {
  title: '',
  children: null,
  show: false,
};

export default SideDrawer;
