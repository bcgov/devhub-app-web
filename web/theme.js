/*
Copyright 2019 Province of British Columbia

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at 

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Created by Patrick Simonian
*/

/*
Emotion theming object

We are transitioning from common elements such as colors, background colors being abstracted away into
a themes provider. This is a very common practice now a days and makes it maintain themes on the fly
docs on theming is here https://emotion.sh/docs/theming#___gatsby

Docs on emotion https://emotion.sh/docs/introduction

--- our idea is to introduce emotion early and slowly replace css modules with emotion styled components
when and as needed. This is a slow processes but will lead to better and more maintaible styling in the long run.
*/

import { SPACING, DEVHUB_PALETTE } from './src/constants/designTokens';
export default {
  colors: {
    ...DEVHUB_PALETTE,
  },
  padding: {
    main: {
      mobile: `${SPACING['2x']} ${SPACING['2x']}`,
      desktop: `${SPACING['2x']} ${SPACING['13x']}`,
    },
  },
  breakpoints: {
    // specific breakpoints
    main: {
      desktop: '@media screen and (min-width: 932px)',
    },
  },
};
