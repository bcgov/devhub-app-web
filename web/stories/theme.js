import { create } from '@storybook/theming';
import { DEVHUB_PALETTE } from '../theme';

export default create({
  base: 'light',

  colorPrimary: 'hotpink',
  colorSecondary: 'deepskyblue',

  // UI
  appBg: 'white',
  appContentBg: 'white',
  appBorderColor: 'grey',
  appBorderRadius: 4,

  // Typography
  fontBase: 'sans-serif',
  fontCode: 'monospace',

  // Text colors
  textColor: DEVHUB_PALETTE.darkgrey,
  textInverseColor: 'rgba(255,255,255,0.9)',

  // Toolbar default and active colors
  barTextColor: DEVHUB_PALETTE.white,
  barSelectedColor: 'black',
  barBg: DEVHUB_PALETTE.blue,

  // Form colors
  inputBg: 'white',
  inputBorder: 'silver',
  inputTextColor: 'black',
  inputBorderRadius: 4,

  brandTitle: 'My custom storybook',
  brandImage:
    'https://u58tqm22fk-flywheel.netdna-ssl.com/wp-content/uploads/2016/06/BC-Gov-Logo-CMYK_pos.png',
});
