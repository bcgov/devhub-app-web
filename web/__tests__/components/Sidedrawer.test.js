import React from 'react';
import { ThemeProvider } from '@emotion/react';
import theme from '../../theme';
import { render, cleanup, fireEvent } from '@testing-library/react';
import SideDrawer, { TEST_IDS } from '../../src/components/SideDrawer/SideDrawer';

describe('Sidedrawer Page', () => {
  let title = 'Foo';
  let show = true;
  let closeDrawer = jest.fn();
  afterEach(cleanup);
  beforeEach(() => {
    closeDrawer.mockReset();
  });

  test('it matches snapshot', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <SideDrawer show={show} title={title} closeDrawer={closeDrawer} />
      </ThemeProvider>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('it calls drawer close fn when x button is clicked', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <SideDrawer show={show} title={title} closeDrawer={closeDrawer} />
      </ThemeProvider>,
    );

    const button = getByTestId(TEST_IDS.closeButton);

    fireEvent.click(button);
    expect(closeDrawer).toHaveBeenCalled();
  });

  test('it calls drawer close fn when backdrop is clicked', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <SideDrawer show={show} title={title} closeDrawer={closeDrawer} />
      </ThemeProvider>,
    );

    const backdrop = getByTestId(TEST_IDS.backdrop);

    fireEvent.click(backdrop);
    expect(closeDrawer).toHaveBeenCalled();
  });
});
