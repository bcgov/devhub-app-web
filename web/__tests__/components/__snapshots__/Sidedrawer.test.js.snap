// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidedrawer Page it matches snapshot 1`] = `
.emotion-0 {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  z-index: 250;
  background-color: rgba(0,0,0,0.35);
}

<div
  class="emotion-0 emotion-1"
  data-testid="side-drawer-backdrop"
/>
`;
